components/brave_wallet/browser/eth_tx_controller.cc

```diff
@@ -0,0 +1,138 @@
+          nonce_tracker_.get())),
+      weak_factory_(this) {}
+
+EthTxController::~EthTxController() = default;
+
+void EthTxController::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void EthTxController::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void EthTxController::AddUnapprovedTransaction(const EthTransaction& tx,
+                                               const EthAddress& from) {
+  EthTxStateManager::TxMeta meta(tx);
+  meta.id = EthTxStateManager::GenerateMetaID();
+  // TODO(darkdh): estimate gas price and limit
+  meta.tx.set_gas_price(20);
+  meta.tx.set_gas_limit(21000);
+  meta.from = from;
+  meta.created_time = base::Time::Now();
+  meta.status = EthTxStateManager::TransactionStatus::UNAPPROVED;
+  tx_state_manager_->AddOrUpdateTx(meta);
+
+  for (Observer& observer : observers_)
+    observer.OnNewUnapprovedTx(meta);
+}
+
+bool EthTxController::ApproveTransaction(const std::string& tx_meta_id) {
+  EthTxStateManager::TxMeta meta;
+  if (!tx_state_manager_->GetTx(tx_meta_id, &meta)) {
+    LOG(ERROR) << "No transaction found";
+    return false;
+  }
+  if (!meta.last_gas_price) {
+    nonce_tracker_->GetNextNonce(
+        meta.from, base::BindOnce(&EthTxController::OnGetNextNonce,
+                                  weak_factory_.GetWeakPtr(), std::move(meta)));
+  } else {
+    OnGetNextNonce(EthTxStateManager::TxMeta(meta), true, meta.tx.nonce());
+  }
+
+  return true;
+}
+
+void EthTxController::RejectTransaction(const std::string& tx_meta_id) {
+  EthTxStateManager::TxMeta meta;
+  if (!tx_state_manager_->GetTx(tx_meta_id, &meta)) {
+    LOG(ERROR) << "No transaction found";
+    return;
+  }
+  meta.status = EthTxStateManager::TransactionStatus::REJECTED;
+  tx_state_manager_->AddOrUpdateTx(meta);
+}
+
+void EthTxController::OnGetNextNonce(EthTxStateManager::TxMeta meta,
+                                     bool success,
+                                     uint256_t nonce) {
+  if (!success) {
+    // TODO(darkdh): Notify observers
+    LOG(ERROR) << "GetNextNonce failed";
+    return;
+  }
+  meta.tx.set_nonce(nonce);
+  auto* keyring_controller = wallet_service_->keyring_controller();
+  DCHECK(!keyring_controller->IsLocked());
+  auto* default_keyring = keyring_controller->GetDefaultKeyring();
+  DCHECK(default_keyring);
+  default_keyring->SignTransaction(meta.from.ToChecksumAddress(), &meta.tx);
+  meta.status = EthTxStateManager::TransactionStatus::APPROVED;
+  tx_state_manager_->AddOrUpdateTx(meta);
+  PublishTransaction(meta.tx, meta.id);
+}
+
+void EthTxController::PublishTransaction(const EthTransaction& tx,
+                                         const std::string& tx_meta_id) {
+  if (!tx.IsSigned()) {
+    LOG(ERROR) << "Transaction must be signed first";
+    return;
+  }
+  EthJsonRpcController* rpc_controller = wallet_service_->rpc_controller();
+  DCHECK(rpc_controller);
+
+  rpc_controller->SendRawTransaction(
```

Comment:
since `eth_tx_controller` depends on `rpc_controller` it should be listed after it in `BraveWalletService` so it always gets destroyed first. I think we're obscuring some potential issues here by passing in the wallet service as mentioned above.

Reply:
issue opened in  brave/brave-browser#16823